nacc_form_validator/nacc_validator.py

```diff
@@ -690,62 +690,58 @@ def _validate_function(self, function: str, field: str, value: object):
             self.__add_system_error(field, err_msg)
             raise ValidationException(err_msg)
 
-    def _check_with_gds(self, field: str, value: object):
+    def _validate_compute_gds(self, keys: List[str], field: str, value: object):
         """Validate Geriatric Depression Scale (GDS) calculation.
 
         Args:
+            keys: GDS computation fields specified in the rule definition
             field: Variable name
             value: Variable value
+
+        Note: Don't remove below docstring,
+        Cerberus uses it to validate the schema definition.
+
+        The rule's arguments are validated against this schema:
+            {
+                'type': 'list', 
+                'minlength': 15,
+                'schema': {'type': 'string'}
+            }
         """
 
         nogds = 0
         if "nogds" in self.document:
             nogds = self.document["nogds"]
 
-        if nogds == 1 and value != 88:
-            self._error(field, ErrorDefs.CHECK_GDS_1)
-            return
-
-        keys = [
-            "satis",
-            "dropact",
-            "empty",
-            "bored",
-            "spirits",
-            "afraid",
-            "happy",
-            "helpless",
-            "stayhome",
-            "memprob",
-            "wondrful",
-            "wrthless",
-            "energy",
-            "hopeless",
-            "better",
-        ]
-
         num_valid = 0
         gds = 0
         for key in keys:
             if key in self.document and self.document[key] in [1, 0]:
                 num_valid += 1
                 gds += self.document[key]
 
-        if nogds == 1 and num_valid >= 12:
-            self._error(field, ErrorDefs.CHECK_GDS_2)
+        if nogds == 1:
+            if num_valid >= 12:
+                self._error(field, ErrorDefs.CHECK_GDS_1, 0)
+
+            if value != 88:
+                self._error(field, ErrorDefs.CHECK_GDS_2, 1)
+
             return
 
-        if nogds == 0 and num_valid < 12:
-            self._error(field, ErrorDefs.CHECK_GDS_3)
+        if not nogds and num_valid < 12:
+            self._error(field, ErrorDefs.CHECK_GDS_3, 2)
             return
 
         error_def = ErrorDefs.CHECK_GDS_4
+        rule_no = 3
         if 0 < num_valid < 15:
```

Comment:
Shouldn't this be if 12 < num_valid < 15 (or 15 - num_valid <= 3), since the rule is **up to 3 of the 15 items are 9 (unanswered)**?